opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -6,139 +6,254 @@
    */
-  public static ExprType getExprType(String openSearchType) {
-    if (OPENSEARCH_TYPE_TO_EXPR_TYPE_MAPPING.containsKey(openSearchType)) {
-      return OPENSEARCH_TYPE_TO_EXPR_TYPE_MAPPING.get(openSearchType);
+  public static OpenSearchDataType of(MappingType mappingType,
+                                      Map<String, OpenSearchDataType> properties,
+                                      Map<String, OpenSearchDataType> fields) {
+    var res = of(mappingType);
+    if (!properties.isEmpty() || !fields.isEmpty()) {
+      // Clone to avoid changing the singleton instance.
+      res = res.cloneEmpty();
+      res.properties = ImmutableMap.copyOf(properties);
+      res.fields = ImmutableMap.copyOf(fields);
     }
-    return EXTRA_OPENSEARCH_TYPE_TO_EXPR_TYPE_MAPPING.getOrDefault(openSearchType, UNKNOWN);
+    return res;
+  }
+
+  protected OpenSearchDataType(MappingType mappingType) {
+    this.mappingType = mappingType;
   }
 
   /**
-   * Convert ExprType to OpenSearch type string.
-   * @param type expr type
-   * @return OpenSearch type string
+   * A constructor function which builds proper `OpenSearchDataType` for given {@link ExprType}.
+   * @param type A type.
+   * @return An instance of `OpenSearchDataType`.
    */
-  public static String getOpenSearchType(ExprType type) {
-    return OPENSEARCH_TYPE_TO_EXPR_TYPE_MAPPING.inverse().get(type);
+  public static OpenSearchDataType of(ExprType type) {
+    if (type instanceof OpenSearchDataType) {
+      return (OpenSearchDataType) type;
+    }
+    var res = instances.getOrDefault(type.toString(), null);
+    if (res != null) {
+      return res;
+    }
+    res = new OpenSearchDataType((ExprCoreType) type);
+    instances.put(type.toString(), res);
+    return res;
   }
 
-  @Override
-  public List<ExprType> getParent() {
-    return parents;
+  protected OpenSearchDataType(ExprCoreType type) {
+    this.exprCoreType = type;
+  }
+
+  protected OpenSearchDataType() {
   }
 
+  // For datatypes with properties (example: object and nested types)
+  // a read-only collection
+  @Getter
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> properties = ImmutableMap.of();
+
+  // text could have fields
+  // a read-only collection
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> fields = ImmutableMap.of();
+
   @Override
+  // Called when building TypeEnvironment and when serializing PPL response
   public String typeName() {
-    return jdbcType;
+    // To avoid breaking changes return `string` for `typeName` call (PPL) and `text` for
+    // `legacyTypeName` call (SQL). See more: https://github.com/opensearch-project/sql/issues/1296
+    if (legacyTypeName().equals("TEXT")) {
+      return "STRING";
+    }
+    return legacyTypeName();
   }
 
   @Override
+  // Called when serializing SQL response
   public String legacyTypeName() {
-    return LEGACY_TYPE_NAME_MAPPING.getOrDefault(this, typeName());
+    if (mappingType == null) {
```

Comment:
in which case mappingType is null?

Reply:
When object created from aggregation, see [changes in AggregationQueryBuilder](https://github.com/opensearch-project/sql/pull/1314/files#diff-56884001cbeae22c3bf07df1ef8d53f7f2fcf51c16a7982db6a1a6d4e940221a). Aggregation adds a new column which type has no index mapping; the column type should be passed into `OpenSearchExprValueFactory` as `OpenSearchDataType` to provide proper data parsing.